### PR TITLE
feat(processor): send line items to Adyen for capture payments if payment-method requires it

### PR DESCRIPTION
**Pull request by joey-koster-ct**
**Ticket [SCC-2373](https://commercetools.atlassian.net/browse/SCC-2373)**

**Description**

For certain payment methods, in this case for `klarna` (Pay Later), `klarna_account` (Pay Over Time) and `klarna_paynow` (Pay Now) it requires us to send the line items to Adyen again during a `capture` API call. Same as during the authorisation call. In this PR we extend the `capture` API code to include this.

Sending lineitems to Adyen is conditional. Only if the payment method is (at the time of this PR) one of the three klarna payment methods will we trigger this new logic.

When a payment capture request is coming in we only know the `paymentID` so via this ID we will fetch the CoCo order and as a fallback the cart. If neither could be found an error is returned to the API caller without the capture request to Adyen happening.

**Technical notes**

In order to be able to fetch the order/cart this PR builds upon the change from https://github.com/commercetools/connect-payments-sdk/pull/93 and the new released version of `0.8.0` from the `connect-payments-sdk` package.

### CODE REVIEW

File: processor/src/services/adyen-payment.service.ts [hunk 1]
```diff
@@ -73,9 +83,7 @@ export class AdyenPaymentService extends AbstractPaymentService {
   private refundPaymentConverter: RefundPaymentConverter;
 
   constructor(opts: AdyenPaymentServiceOptions) {
-    super(opts.ctCartService, opts.ctPaymentService);
-    this.ctCartService = opts.ctCartService;
-    this.ctPaymentService = opts.ctPaymentService;
```
Comment (joey-koster-ct) [inline processor/src/services/adyen-payment.service.ts:78]: There is no need to set these values manually. The `AbstractPaymentService` already does this.

File: processor/src/services/adyen-payment.service.ts [hunk 2]
```diff
@@ -397,11 +405,20 @@ export class AdyenPaymentService extends AbstractPaymentService {
...
 
   async capturePayment(request: CapturePaymentRequest): Promise<PaymentProviderModificationResponse> {
+    // Only process if the payment method requires it so we don't fetch order/cart unnecessary.
+    let adyenLineItems: LineItem[] | undefined = undefined;
+    if (
+      request.payment.paymentMethodInfo.method &&
+      METHODS_REQUIRE_LINE_ITEMS.includes(request.payment.paymentMethodInfo.method)
+    ) {
+      adyenLineItems = await this.retrievePaymentAssociatedLineItems(request.payment.id);
+    }
```
Comment (joey-koster-ct) [inline processor/src/services/adyen-payment.service.ts:415]: There are multiple ways of doing the logic and retrieving the order/cart and then extracting the Adyen line items from it. I chose this way cause the `AdyenApi` supports sending either a `LineItem[] | undefined` for the `lineItem` property in the payload to the `captureAuthorisedPayment` via the converter.

This way the properties line up nicely.

File: processor/src/services/converters/capture-payment.converter.ts [hunk 3]
```diff
@@ -1,16 +1,18 @@
 import { config } from '../../config/config';
 import { PaymentCaptureRequest } from '@adyen/api-library/lib/src/typings/checkout/paymentCaptureRequest';
 import { CapturePaymentRequest } from '../types/operation.type';
+import { LineItem } from '@adyen/api-library/lib/src/typings/checkout/lineItem';
 
 export class CapturePaymentConverter {
-  public convertRequest(opts: CapturePaymentRequest): PaymentCaptureRequest {
+  public convertRequest(opts: CapturePaymentRequest, lineItems: LineItem[] | undefined): PaymentCaptureRequest {
```
Comment (joey-koster-ct) [inline processor/src/services/converters/capture-payment.converter.ts:7]: Type of `LineItem[] | undefined` is exactly what the Adyen `PaymentCaptureRequest` accepts.

File: processor/test/services/adyen-payment.service.spec.ts [hunk 4]
```diff
@@ -42,17 +46,11 @@ import { PaymentResponse } from '@adyen/api-library/lib/src/typings/checkout/pay
...
-import { ApplePayPaymentSessionError } from '../../src/errors/adyen-api.error';
 
 interface FlexibleConfig {
   [key: string]: string; // Adjust the type according to your config values
 }
 
-export type AdyenPaymentServiceOptions = {
-  ctCartService: CommercetoolsCartService;
-  ctPaymentService: CommercetoolsPaymentService;
-};
```
Comment (joey-koster-ct) [inline processor/test/services/adyen-payment.service.spec.ts:54]: We redeclared the service options which imo was not needed caus we already have that in the `adyen-payment.service` so I removed it from here.

File: processor/src/services/converters/capture-payment.converter.ts [hunk 5]
```diff
@@ -1,16 +1,74 @@
...
+  private ctCartService: CommercetoolsCartService;
+  private ctOrderService: CommercetoolsOrderService;
+
+  constructor(ctCartService: CommercetoolsCartService, ctOrderService: CommercetoolsOrderService) {
+    this.ctCartService = ctCartService;
+    this.ctOrderService = ctOrderService;
+  }
+
+  public async convertRequest(opts: CapturePaymentRequest): Promise<PaymentCaptureRequest> {
+    // Only process if the payment method requires it so we don't fetch order/cart unnecessary.
```
Comment (janmooij) [inline processor/src/services/converters/capture-payment.converter.ts:28]:
```suggestion
    // Only process if the payment method requires it so we don't fetch order/cart unnecessarily.
```

**Review by janmooij — APPROVED** [review]
lgtm 👍

**Review by dasanorct — APPROVED** [review]

**Pull request merged**